presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -120,24 +137,63 @@ public static CachingHiveMetastore memoizeMetastore(ExtendedHiveMetastore delega
                 newDirectExecutorService(),
                 OptionalLong.empty(),
                 OptionalLong.empty(),
-                maximumSize);
+                maximumSize,
+                false,
+                ALL);
     }
 
-    private CachingHiveMetastore(ExtendedHiveMetastore delegate, ExecutorService executor, OptionalLong expiresAfterWriteMillis, OptionalLong refreshMills, long maximumSize)
+    private CachingHiveMetastore(
+            ExtendedHiveMetastore delegate,
+            ExecutorService executor,
+            OptionalLong expiresAfterWriteMillis,
+            OptionalLong refreshMills,
+            long maximumSize,
+            boolean partitionVersioningEnabled,
+            MetastoreCacheScope metastoreCacheScope)
     {
         this.delegate = requireNonNull(delegate, "delegate is null");
         requireNonNull(executor, "executor is null");
-
-        databaseNamesCache = newCacheBuilder(expiresAfterWriteMillis, refreshMills, maximumSize)
+        this.partitionVersioningEnabled = partitionVersioningEnabled;
+
+        OptionalLong cacheExpiresAfterWriteMillis;
+        OptionalLong cacheRefreshMills;
+        long cacheMaxSize;
+
+        OptionalLong partitionCacheExpiresAfterWriteMillis;
+        OptionalLong partitionCacheRefreshMills;
+        long partitionCacheMaxSize;
+
+        switch (metastoreCacheScope) {
+            case PARTITION:
+                partitionCacheExpiresAfterWriteMillis = expiresAfterWriteMillis;
+                partitionCacheRefreshMills = refreshMills;
+                partitionCacheMaxSize = maximumSize;
+                cacheExpiresAfterWriteMillis = OptionalLong.of(0);
+                cacheRefreshMills = OptionalLong.of(0);
+                cacheMaxSize = 0;
+                break;
+
+            case ALL:
+            default:
```

Comment:
nit: how about moving the following assignments under `case ALL`, and throw for `default`?